### DISCUSSION TITLE
How to redirect all google search to a different search engine

### DISCUSSION THREAD

**Discussion by m-zheng** [mshibanami/redirect-web#4] [Questions] [answered]
I am wondering if it is possible to create a regular expression based rule to redirect all google search to a different search engine that is not supported in Safari. This requires the rule to capture the value of the parameter `q` in the google search url.

I have created a wild card based rule as shown below. It will convert `https://www.google.com/search?client=safari&rls=en&q=what+is+my+ip+address&ie=UTF-8&oe=UTF-8&safe=active` to `https://www.startpage.com/do/dsearch?query=what+is+my+ip+address`. 

```json
{
  "appVersion" : "4.5.0",
  "bundleID" : "io.github.mshibanami.RedirectWebForSafari",
  "createdAt" : "2023-03-06T17:03:53.892Z",
  "kind" : "RedirectList",
  "redirects" : [
    {
      "appOpenerPostActions" : [

      ],
      "captureGroupProcesses" : [

      ],
      "comments" : "",
      "destinationURLPattern" : "https:\/\/www.startpage.com\/do\/dsearch?query=$1",
      "exampleURLs" : [
        "https:\/\/www.google.com\/search?client=safari&rls=en&q=what+is+my+ip+address&ie=UTF-8&oe=UTF-8&safe=active"
      ],
      "excludeURLPatterns" : [

      ],
      "isEnabled" : true,
      "kind" : "Redirect",
      "sourceURLPattern" : {
        "type" : "wildcard",
        "value" : "https:\/\/www.google.com\/search?client=safari&rls=en&q=*&ie=UTF-8&oe=UTF-8&safe=active"
      },
      "title" : "Redirect google search to startpage"
    }
  ]
}
```

However, the rule above will not work on a different apple device because Safari appends different parameters to the url based on os. For example, Safari in ipad creates `https://www.google.co.uk/search?q=what+is+my+ip+address&ie=UTF-8&oe=UTF-8&hl=en-gb&client=safari`. This has become a problem because rules are sync in icloud across multiple apple devices.

I am really struggling to create a regular expression based rule to redirect all google search to a different search engine. Any help would be highly appreciated!

**Comment by mshibanami** [comment, marked as answer]
Hi @m-zheng, welcome to this forum! You're the first visitor. 🙂 And sorry for any inconvenience. I agree that parsing a Google Search URL is confusing, and I think using regular expressions is the best way to achieve your goal.

If all you need for startpage.com is `q`, I think you can try this rule:

- Redirect From:
  - URL Pattern: `https://www\.google\.com/search\?(.+&)?q=([^&]+)(&.*)?`
      - If you use `google.co.uk` as well, you would prefer `https://www\.google\.(?:com|co.uk)/search\?(.+&)?q=([^&]+)(&.*)?` 
  - Pattern Type: `Regular Expression`
- Redirect To:
  - `https://www.startpage.com/do/dsearch?query=$2`

I hope it helps.

![](https://user-images.githubusercontent.com/1333214/223283369-b4600fc1-5361-49e4-b1db-7888f0d67491.png)

**Comment by m-zheng** [comment]
Thank you very much for your help! This rule can simplify all wild card based rules, which is superb! I ended using the one below for my firefox as well.
`^(?:https?:\/\/)(?:www.)google\.(?:com|au|de|co.uk)\/search\?(.+&)?q=([^&]+)(&.*)?`

**Reply by mshibanami** [reply]
Great! Right, it's more flexible. Thanks for sharing!

**Reply by m-zheng** [reply]
This helps me override the default search engine to any search engine on macos globally, which is never allowed by Apple. Thank you again!
